File: stacks/problems/next_greater_element/NextGreaterElement.java
import java.util.Arrays;
import java.util.EmptyStackException;

public class Stack {
  // NOTE: top refers to index of array
  private int top;
  private int[] arr;

  // Constructors
  public void push(int data) {
    if (isFull()) {
      throw new RuntimeException("Stack Is Full !!!");
    }
    top++;
    arr[top] = data;
  }

  public int pop() {
    if (isEmpty()) {
      throw new RuntimeException("Stack Is Empty !!!");
    }
    int result = arr[top];
    top--;
    return result;
  }

  public int peek() {
    if (isEmpty()) {
      throw new RuntimeException("Stack Is Empty !!!");
    }
    return arr[top];
  }

  public boolean isFull() { return arr.length == size(); }
  public boolean isEmpty() { return top < 0; }
  public int size() { return top + 1; }

  public Stack(int capacity) {
    top = -1;
    arr = new int[capacity];
  }

  public Stack() { this(10); }

  public int[] nextGreaterElement(int[] arr) {
    int[] result = new int[arr.length];
    for (int i = arr.length - 1; i >= 0; i--) {
      if (!stack.isEmpty()) {
        while (!stack.isEmpty && stack.peek() <= arr[i]) {
          stack.pop();
        }
      }
      if (stack.isEmpty()) {
        result[i] = -1;
      } else {
        result[i] = stack.peek();
      }
      stack.push(arr[i]);
    }
    return result;
  }

  public static void main(String[] args) {
    Stack stack = new Stack();
    stack.push(4);
    stack.push(7);
    stack.push(3);
    stack.push(4);
    stack.push(8);
    stack.push(1);
    System.out.println(Arrays.toString(stack));
  }
}

File: stacks/problems/next_greater_element/NextGreaterElement.js
// NOTE: Tutorial doesn't provide proper context for
// establishing this in java and I'm lazy, so here it is in JS:

/*
 * A somewhat complicated algorithm for a beginner.
 * Here we traverse our original array backwards,
 * pushing the first number to be higher than the number at arr[i]
 * onto the stack.
 * If, however, the latest addition to the stack (the top)
 * is less than the inspected number at arr[i], the value is popped off the
 * stack and the while loop continues.
 *
 * Once that terminates, the stack is checked to see if it is empty,
 * if it is, then the result[i] is stored a value of -1,
 * otherwise, the top element of the stack is assigned to result[i].
 *
 * Finally the latest element of the array is pushed onto the
 * stack to be compared again in the while loop above. This is done until
 * all elements are traversed backwards through the original arr
 * */

// initial array = [4, 7, 3, 4, 8, 1]
const nextGreaterElement = (arr = []) => {
  const result = [];
  const stack = [];
  for (let i = arr.length - 1; i >= 0; i--) {
    if (stack.length > 0) {
      while (stack.length > 0 && stack[stack.length - 1] <= arr[i]) {
        stack.pop();
      }
    }
    if (!stack.length) {
      result[i] = -1;
    } else {
      result[i] = stack[stack.length - 1];
    }
    stack.push(arr[i]);
  }
  return result;
};

const resultArr = nextGreaterElement([4, 7, 3, 4, 8, 1]);

console.log("resultArr :=>", resultArr);
// resultArr :=> [ 7, 8, 4, 8, -1, -1 ]
